### PR TITLE
Vector serialization with separator

### PR DESCRIPTION
**Pull request by thomas-lemaire**
This PR changes the behavior of helper::vector to make it closer to svector
- helper::vector can parse space separated values **and** [,] separated values
- helper::vector now serialize into [,] separated string
- helper::fixed_array (and so helper::Vec<Type>) can also parse space separated values **and** [,] separated values

As a consequence the serialization/deserialization on the python side (SofaPython plugin) is updated accordingly. Now the repr() of a python object is used to serialize it into a string in the Node.createObject() method if it is not a string already.

Later on, code can be cleaned-up:
- no more cat/concat/listToStr() calls needed in python code
- no more helper::svector needed in c++ code

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: applications/plugins/SofaPython/Binding_BaseContext.cpp [hunk 1]
```diff
@@ -92,37 +92,6 @@ static std::ostream& pythonToSofaDataString(PyObject* value, std::ostream& out)
         return out;
     }
 
-    if( PySequence_Check(value) )
```
Comment (damienmarchal) [inline applications/plugins/SofaPython/Binding_BaseContext.cpp:95]: There was a long discussion in which we decided not to rely only on "repr/str" and instead allow objects to implement their own conversion operator to pass to the create function. Have a look for "getAsACreateObjectParameter" in Binding_BaseContext.cpp.

Removing the call as you do here suppress this behavior.

File: applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp [hunk 2]
```diff
@@ -156,13 +156,13 @@ std::vector<std::vector<std::string>> dataconversionvalues =
...
-     {"[['aa','bb'], ['cc','dd'], ['ee','ff']]", "aa bb cc dd ee ff"},
-     {"range(1,5)", "1 2 3 4"},
-     {"xrange(1,5)", "1 2 3 4"},
+     {"[1, 2, 3, 4]", "[1, 2, 3, 4]"},
+     {"[1.0, 2.0, 3.0, 4.0]", "[1.0, 2.0, 3.0, 4.0]"},
+     {"['ab', 'cd', 'ef', 'gh']", "['ab', 'cd', 'ef', 'gh']"},
+     {"[[1,2], [3,4], [5,6]]", "[[1, 2], [3, 4], [5, 6]]"},
+     {"[['aa','bb'], ['cc','dd'], ['ee','ff']]", "[['aa', 'bb'], ['cc', 'dd'], ['ee', 'ff']]"},
+     {"range(1,5)", "[1, 2, 3, 4]"},
+     {"xrange(1,5)", "xrange(1, 5)"},
```
Comment (damienmarchal) [inline applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp:165]: I'm not sure..shouldn't the xrange(1, 5) not returning [1,2,3,4] ?

**Comment by damienmarchal** [comment]
Hi Thomas,

Currently it seems there is some problem in compiling it, but fundamentally I think this is a great objective. 
Thank you very much to tackling this. 

If I understand correctly now we can have the two syntax to read vectors (the one from Python and the historical one). +Backward compatibility is preserved and you added tests to validate the new behavior. This all looks great to me. 

You also changed the getStringValue...here I think this is a problem as it will break a lot of third party code. This is why you had to fix a lot of failing test. Instead of changing the behavior of an existing function it would be much better to offer a second function to serialize to python format (or adds a parameter to the existing one to control the behavior eg: getStringValue(SerializingMode mode=SerializingMode::ClassicSofa)). The underlying idea is that by default old code will function as usual and "awared" users can take profit of the existing function getStringValueAsPython() or getStringValue(SerializingMode::Python). 

I'm sorry because I think it probably took you a lot of time to update the failing tests but the tests shouldn't have been changed, only new one should have been added to validate the new syntax and corner case. 

Minor details: 
There is some msg_error("(S)Vector") in vector shouldn't it be vector<> instead ?. 
Maybe you will find this pattern usefull: 
```cpp
namespace sofa {
     namespace helper {
           namespace logging {
               inline bool notMuted(const sofa::helper::io::MeshVTK* ){ return true; }
              inline ComponentInfo::SPtr getComponentInfo(const sofa::helper::io::MeshVTK*)
              {
                     return ComponentInfo::SPtr(new ComponentInfo("MeshVTK")) ;
              }
       } /// logging
    } /// helper
} /// sofa
```

It allows to write the implicit (this) msg_* as in   
```cpp
msg_error() << "Blahblah"
```

Instead of
```cpp
msg_error("MeshVTK") << "Blahblah"
```

**Review by damienmarchal — CHANGES_REQUESTED** [review]
This PR address a very valid issue but some more work is needed to preserve backward compatibility of getValueString.

**Comment by damienmarchal** [comment]
Beward, you are working in the sofa-consortium repo instead of in the Anatoscope's fork. 
The problem is that it mess things in the CI.

File: applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp (hunk 2 above)
Comment (thomas-lemaire) [reply applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp:165]: xrange() returns an iterator, not a list

File: applications/plugins/SofaPython/Binding_BaseContext.cpp (hunk 1 above)
Comment (thomas-lemaire) [reply applications/plugins/SofaPython/Binding_BaseContext.cpp:95]: Ok, I was not aware of the whole discussion, but from discussion here it looks the right way to me. If a sofa type needs to implement its own python serialization, then it is a matter to implement the __repr__() function for that type. It sounds more standard, doesn't it ?

**Comment by thomas-lemaire** [comment]
As an overall comment many **thanks** for taking time reviewing this PR !

**Comment by thomas-lemaire** [comment]
About the failing tests I had to fix and serialization changes...
*I wanted to reply in the specific comment but could not (github magic effect ??)*

## Failing tests
The failing tests were using the *serialization* to check the result, I mean the tests are doing comparison of **string** instead of the **actual** values, but these tests are not testing the serialization (in which case it would have been correct).

These tests should be changed to compare the expected returned values (with a tolerance in the case of floating point), using the Sofa_test.h EXPECT_VEC_DOUBLE_EQ and other similar macros (add some if they are missing). Not to mix every thing it could be addressed in an other PR.

## serialization using "[,]" instead of " "
As long as the deserialization operator is compatible with the serialization one, I do not see any issue. And no example scene is failing because of that.

**Comment by thomas-lemaire** [comment]
About the repository, I also have a similar PR on the anatoscope repo, this is why I pushed this branch here.

**Comment by thomas-lemaire** [comment]
I see one incompatibility with the existing : when a helper::vector<helper::SVector<Type>> or a helper::SVector<helper::vector<Type>> is used, there is no way to discriminate between them.

**Comment by damienmarchal** [comment]
Hi Thomas,

You said:
"As long as the deserialization operator is compatible with the serialization one, I do not see any issue. And no example scene is failing because of that."

You have to take the whole picture into consideration, everyone is not using serialization/deserialization as "pairs". The actual default serialization format is used in a lot of other software (GUIs, scene processing tools written in python, perl and C++). If you change the behavior of the default serialization mechanism all of them will stop working and required to be updated. This is a very, very, very nasty thing to impose to other users of Sofa. A PR doing has no changes to be accepted. 

Actually tests is part of our guard line because they validates the API compatibility and allow us to detect changes that breaks it. The existing tests should be considered as validating the "the existing" sofa API.  It was normal to have  them to fail on your PR because it shows that the PR was breaking the compatibility. 

I really like the idea of keeping the old serialization behavior as it is and add the other deserialization as an extra feature. This will make the whole PR transparent for users that don't want to update their code base. It thus it can be merged very quickly. 

About the test: 
You are right that it would be better if we had test specific tests to validate the serialization alone.
I personally would like to have tests to validate that getValueString() return "1 2 3" and that getValueAsPythonString() return "[1 2 3]" . And also as you said having tests to validate against the real values instead of their serialization would be nice. . 

About the branches:
The rule is to never work directly on the sofa-framework repository. There is no reason to do so (if you really want to make two PR one in anatoscope and one in sofa ... you need to to that from two branches located in the anatoscope repo). Otherwise this consume a lot of disk space on the CI servers that we have to fix manually. Now fundamentally from my experience in playing with PR and branches I think that making two PR will be very complex for you to handle. Because the PR you issued to anatoscope and the one you issued to sofa-framework may diverge ... so it will cost you to switch from one to the other and handle the changes. 

PS: I'm not very good to explains things in english so if you want we can continue in by email in french.

File: applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp (hunk 2 above)
Comment (damienmarchal) [reply applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp:165]: In the past, xrange was evaluated by the createObject function because it was implementing the Sequence operator. I see no reason why this should change.

**Comment by thomas-lemaire** [comment]
Closing as PR branch on the consortium repository is shouting the CI
